packages/bolt-connection/test/bolt/bolt-protocol-v4x4.test.js

```diff
@@ -0,0 +1,356 @@
+    expect(protocol.flushes).toEqual([true])
+  })
+
+  it('should request routing information', () => {
+    const recorder = new utils.MessageRecordingConnection()
+    const protocol = new BoltProtocolV4x4(recorder, null, false)
+    utils.spyProtocolWrite(protocol)
+    const routingContext = { someContextParam: 'value' }
+    const databaseName = 'name'
+
+    const observer = protocol.requestRoutingInformation({
+      routingContext,
+      databaseName
+    })
+
+    protocol.verifyMessageCount(1)
+    expect(protocol.messages[0]).toBeMessage(
+      RequestMessage.routeV4x4(routingContext, [], { databaseName, impersonatedUser: null })
+    )
+    expect(protocol.observers).toEqual([observer])
+    expect(observer).toEqual(expect.any(RouteObserver))
+    expect(protocol.flushes).toEqual([true])
+  })
+
+  it('should request routing information sending bookmarks', () => {
+    const recorder = new utils.MessageRecordingConnection()
+    const protocol = new BoltProtocolV4x4(recorder, null, false)
+    utils.spyProtocolWrite(protocol)
+    const routingContext = { someContextParam: 'value' }
+    const listOfBookmarks = ['a', 'b', 'c']
+    const bookmark = new Bookmark(listOfBookmarks)
+    const databaseName = 'name'
+
+    const observer = protocol.requestRoutingInformation({
+      routingContext,
+      databaseName,
+      sessionContext: { bookmark }
+    })
+
+    protocol.verifyMessageCount(1)
+    expect(protocol.messages[0]).toBeMessage(
+      RequestMessage.routeV4x4(routingContext, listOfBookmarks, { databaseName, impersonatedUser: null})
+    )
+    expect(protocol.observers).toEqual([observer])
+    expect(observer).toEqual(expect.any(RouteObserver))
+    expect(protocol.flushes).toEqual([true])
+  })
+
+  it('should run a query', () => {
+    const database = 'testdb'
+    const bookmark = new Bookmark([
+      'neo4j:bookmark:v1:tx1',
+      'neo4j:bookmark:v1:tx2'
+    ])
+    const txConfig = new TxConfig({
+      timeout: 5000,
+      metadata: { x: 1, y: 'something' }
+    })
+    const recorder = new utils.MessageRecordingConnection()
+    const protocol = new BoltProtocolV4x4(recorder, null, false)
+    utils.spyProtocolWrite(protocol)
+
+    const query = 'RETURN $x, $y'
+    const parameters = { x: 'x', y: 'y' }
+
+    const observer = protocol.run(query, parameters, {
+      bookmark,
+      txConfig,
+      database,
+      mode: WRITE
+    })
+
+    protocol.verifyMessageCount(2)
+
+    expect(protocol.messages[0]).toBeMessage(
+      RequestMessage.runWithMetadata(query, parameters, {
+        bookmark,
+        txConfig,
+        database,
+        mode: WRITE
+      })
+    )
+    expect(protocol.messages[1]).toBeMessage(RequestMessage.pull())
+    expect(protocol.observers).toEqual([observer, observer])
+    expect(protocol.flushes).toEqual([false, true])
+  })
+
+  it('should run a with impersonated user', () => {
+    const database = 'testdb'
+    const impersonatedUser = 'the impostor'
```

Comment:
```suggestion
    const impersonatedUser = 'the imposter'
```

Not that it really matters :rofl: (it's also misspelled in other places).

Reply:
I will keep it since the imposto is imposter in portuguese :D

Reply:
:brazil: